### ISSUE TITLE
About memory leak of controller.iterate

### ISSUE THREAD

**Issue by WooSunghyeon** [LiuXiaoxuanPKU/GACT-ICML#5]
Hi,

i'm very interested in your wonderful idea of GACT, and therefore i performed GACT in BERT-Large.

I attained identical memory saving and accuracy to that of manuscripts.

However, when i performed experiments enlarging maximum_layer, width, and intermediate_size, FP32 can enlarge model size more than GACT-2bit.

I found that it is because controller.iterate(backprop). With this comments, the memory leak exponentially occurs (not torch.cuda.allocated_memory, but real gpu allocation which can bee seen by nvidia-smi).

Because this comments is essential idea of GACT, we cannot delete this comments.

How can i solve this problem? I want to know how can you attain Table 4 of manuscripts.

I found that sample_grad in **/GACT-ICML/gact/gact/autoprec.py, 
grad.append(param.grad.ravel()) save all param.grad**, and i think it occurs cache memory leak.

**Comment by LiuXiaoxuanPKU** [comment]
Hi thanks for your interest!! Maybe get a bit more information here:
1. how did you enlarging maximum_layer, width, and intermediate_size? I produced the graph using this [file](https://github.com/LiuXiaoxuanPKU/GACT-ICML/blob/release_v6/benchmark/text_classification/exp_mem_speed.py). It's just `python exp_mem_speed.py --mode binary_search_max_hidden_size'.
2. did you disable logging but setting `log_interval=-1` [here](https://github.com/LiuXiaoxuanPKU/GACT-ICML/blob/5ddd88357bcf7ed8341a9b6a7be4a1ac4cdf88a9/gact/gact/conf.py#L63).

**Comment by WooSunghyeon** [comment]
Thank you for your reply.

In my experiments, i produced my results by same code with you (python exp_mem_speed.py --mode binary_search_max_hidden_size and log_interval=-1). But in my experiments, FP32 can attain larger models than GACT-2bit in RTX-3090Ti.

I think this situation is related with actual GPU memory allocation.

When i performed

```
python run_glue.py --model_name_or_path bert-large-cased --task_name mnli --max_length 128 --per_device_train_batch_size 16 --per_device_eval_batch_size 128 --learning_rate 1e-5 --num_train_epochs 10 --seed 43 --pad_to_max_length  --output_dir log/sst2/L2/ 
```
for FP32, and the actual GPU memory allocation is 11818MB by nvidia-smi.

we also performed
```
python run_glue.py --model_name_or_path bert-large-cased --task_name mnli --max_length 128 --per_device_train_batch_size 16 --per_device_eval_batch_size 128 --learning_rate 1e-5 --num_train_epochs 10 --seed 43 --pad_to_max_length  --output_dir log/sst2/L2/   --opt_level L2.2 --gact 
```
for gact 2-bit, and the actual GPU memory allocation is 13975M by nvidia-smi.

Although the memory allocated by Tensor is small than FP32, GACT actually allocates more memory to GPU in my experiments.

In my opinion, this situations occur when calculating sensitivity. When calculating sensitivity, we have to save all parameters of gradient (as implemented ```sample grad``` in https://github.com/LiuXiaoxuanPKU/GACT-ICML/blob/release_v6/gact/gact/autoprec.py)
and it caused memory leak.

**Comment by WooSunghyeon** [comment]
I found that this problem can be solved by changing algorithm. Do not store all gradient of parameters, but store all normalization of gradient of parameters.

I changed ```sample grad``` in https://github.com/LiuXiaoxuanPKU/GACT-ICML/blob/release_v6/gact/gact/autoprec.py  from
```ruby
def sample_grad():
            grad = []
            for param in self.model.parameters():
                if param.grad is not None:
                    grad.append(param.grad.ravel())
            return torch.cat(grad,0)
```
to
```ruby
def sample_grad():
            grad = []
            for param in self.model.parameters():
                if param.grad is not None:
                    grad.append(param.grad.norm().unsqueeze(0))
            return torch.cat(grad,0)
```

and then the GPU memory allocation by nvidia-smi is decreased to 7805MB.

Also, this modified algorithm can attain similar accuracy to original gact.

**Comment by LiuXiaoxuanPKU** [comment]
Great!! Thanks for explaining your findings/fix!
